docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Introduction to Contexts and Dependency Injection
+
+include::./attributes.adoc[]
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+:toc:
+
+In this guide we're going to describe the basic principles of the programming model used all over the Quarkus universe. 
+Quarkus programming model is based on the http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html[Contexts and Dependency Injection for Java 2.0, window="_blank"] specification.
+However, a full CDI implementation is not used under the hood.
+Quarkus only implements a subset of the CDI features - see also <<cdi-reference.adoc#supported_features,the list of supported features>> and <<cdi-reference.adoc#limitations,the list of limitations>>.
+On the other hand, there are quite a few <<cdi-reference#nonstandard_features,non-standard features>> and <<cdi-reference.adoc#build_time_apis,Quarkus-specific APIs>>.
+We believe that our efforts will drive the innovation of the CDI specification towards the build-time oriented developer stacks in the future.
+
+== _Q: OK. Let's start simple. What is a bean?_
+
+A: Well, a bean is a _container-managed_ object that supports a set of basic services, such as injection of dependencies, lifecycle callbacks and interceptors.
+
+== _Q: Wait a minute. What does it mean container-managed?_
+
+A: Simply put, you don't control the lifecycle of the object instance directly.
+Instead, you can affect the lifecycle through the declarative means, such as annotations, configuration, etc.
+The container is the _environment_ where your application runs.
+It creates and destroys the instances of beans, associates the instances with a designated context, and injects them into other beans.
+
+== _Q: What is it good for?_
+
+A: An application developer can focus on the business logic rather than "where and how" to obtain a fully initialized component with all of its dependencies.
+
+NOTE: You've probably heard of the _inversion of control_ (IoC) programming principle. Dependency injection is one of the implementation techniques of IoC. 
+
+== _Q: OK. And what does a bean look like?_
+
+A: There are several kinds of beans.
+The most common ones are class-based beans:
+
+.Simple Bean Example
+[source,java]
+----
+import javax.inject.Inject;
+import javax.enterprise.context.ApplicationScoped;
+import org.eclipse.microprofile.metrics.annotation.Counted;
+
+@ApplicationScoped <1>
+public class Translator {
+
+    @Inject
+    Dictionary dictionary; <2>
+    
+    @Counted  <3>
+    String translate(String sentence) { 
+      // ...
+    }
+}
+----
+<1> This is a scope annotation. It tells the container which context to associate the bean instance with. In this particular case, a *single bean instance* is created for the application and used by all dependants.
```

Comment:
I am sure that users will now what `dependants` means here. Wouldn't it better to say `and used throughout the application`?

Reply:
Yeah, that sounds fine aswell

Reply:
What about _"created for the application and used by all other beans that inject `Translator`."_? It's not so precise but it reads well...